### ISSUE TITLE
[2010-11-01] ComponentSet lacks any kind of onchange triggers

### ISSUE THREAD

**Issue by silverstripe-issues** [silverstripe/silverstripe-framework#1923]
created by: @sminnee (sminnee)
created at: 2010-11-01
original ticket: http://open.silverstripe.org/ticket/6157

---

Right now there is no way to add a trigger that is called when ComponentSets have items added to them, removed, or their extradata modified.

We want some way of adding triggers for these events.  There are a few ways of doing this:
- Call onRelationChanged triggers on the relevant DataObjects
- Subclass ComponentSet and apply the triggers to the subclasses

The former is probably the most appropriate. `onRelationChanged($relationName, $changes)` could be a good general pattern; we would need to find a useful way of summarising the changes that were made to the relation.

**Comment by tractorcow** [comment]
To add: extension points to:

 - add
 - remove
 - setByIDLIst

**Comment by jakxnz** [comment]
I did a quick review of points where a trigger for change might be relevant. I omitted the scope of DataList, so I have listed the inherited method in the subclass scope below. Here are the methods:

RelationList

- forForeignID
- foreignIDFilter

HasManyList

- add
- remove
- removeByID
- removeMany
- removeAll
- setByIDList
- filterByCallback

ManyManyList

- add
- remove
- removeByID
- removeMany
- removeAll
- setByIDList
- filterByCallback

ManyManyThroughList

- add
- remove
- removeByID
- removeMany
- removeAll
- removeByID
- setByIDList
- filterByCallback

While I was reviewing, some questions came to mind:

- Do we want to catch changes when a foreign key is set e.g `ManyManyThroughQueryManipulator::setForeignKey()`?
- Do we want to invoke a trigger every time something like `add()` is called by a higher-level method such as `setByIDList()`? Or do we only want to trigger once in a scenario like that?

**Comment by sminnee** [comment]
Wow, raised in 2010. Sometimes ticket triage feels more like archeology! Thanks for outlining 

To answer your questions, IMO:

"Do we want to catch changes when a foreign key is set e.g ManyManyThroughQueryManipulator::setForeignKey()?"

No. I believe that setForeignKey is more about configuring the relation, so I don't believe it should trigger an onRelationChanged.

"Do we want to invoke a trigger every time something like add() is called by a higher-level method such as setByIDList()? Or do we only want to trigger once in a scenario like that?"

In general I think that there should be ways of tying a single event handler to a 'batch' of changes. From memory, I think a lot of performance issues in, say, composite-object-versioning, are made worse by the lack of such a feature.

However, for some implementations it _may_ be desirable to implement a single-record handler and have the system call it multiple times. I'd want to think that through further, though – it may get messy.

**Comment by jakxnz** [comment]
@sminnee how integral would the relation name (component name) be in the `onRelationChanged($relationName, $changes)`?

I'm guessing it's quite integral because it's the only thing that provides significant context.

The `RelationList`s don't have access to the component name within their existing scope. Assigning it during constructor changes the ORM API, and because DataLists are "immutable" setting the component name via a mutator doesn't have a strong precedent. Also, assigning the relation name for lists created by `DataObject::inferReciprocalComponent()` and `ManyManyThroughQueryManipulator::getParentRelationship()` is less relevant, but leaves some unexpected behaviour from the change trigger if it's ever utilised in that context. And as far as I'm aware, a reverse lookup of the component name is only possible by parsing a `$belongs_to` config.

Basically, having the relation name on hand at the time of the change trigger is not impossible, but not trivial either.

I have enough insight that I could provide a working prototype. Let me know if that would be useful.

**Comment by jakxnz** [comment]
Here's a WIP (still needs unit tests and docs)

https://github.com/silverstripe/silverstripe-framework/compare/4...jakxnz:pulls/4.6/relationlist-triggers

I've used an "observer" pattern (not the SPL pattern), to try and keep the RelationList meta as light as possible. It also means the Injector can govern if the observer is configured, which also makes the overhead optional.

The resulting usage pattern looks something like:

mysite.yml
```
SilverStripe\Core\Injector\Injector:
  SilverStripe\ORM\Observer\RelationListObserver:
    class: Silverstripe\ORM\Observer\RelationListChangeObserver
```

MyDataObject.php
```
class MyDataObject extends DataObject {

    ...

    /**
     * @var array
     */
    private static $has_many = [
        'OtherDataObjects' => OtherDataObject::class,
    ];

    ...

    /**
     * @param $relationName
     * @param $changed
     */
    public function onRelationChanged($relationName, $changed) {
        // $relationName == 'OtherDataObjects'

        // $changed == [
        //   1 => [
        //     'className => 'SilverStripe\ORM\DataObject',
        //     'id' => 1,
        //     'extraFields' => [
        //        'fieldName' => 'value',
        //     ],
        //     'type' => 'added',
        //   ]
        // ]

       ...
    }

    // and/or

    /**
     * @param $changed
     */
    public function onOtherDataObjectsChanged($changed) {
         // $changed == same as above

        ...
    }

}
```

Also, this issue doesn't feel like `effort/easy`, but maybe I'm over-engineering.

Is this in the right direction?
